quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/web/LoadSheddingRequestFilter.java

```diff
@@ -0,0 +1,103 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.quarkus.runtime.integration.web;
+
+import java.util.Optional;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Predicate;
+
+import org.apache.http.HttpStatus;
+import org.jboss.logging.Logger;
+
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+import org.keycloak.quarkus.runtime.configuration.Configuration;
+
+/**
+ * Limit the number of concurrent requests (both blocking and non-blocking) to a configured upper limit.
+ * Special paths (like for health probes and metrics) can be excluded, so their requests will not be dropped.
+ *
+ * @author Alexander Schwartz
+ */
+public class LoadSheddingRequestFilter implements Handler<RoutingContext> {
+
+    private final Logger logger = Logger.getLogger(LoadSheddingRequestFilter.class);
+
+    private final Predicate<RoutingContext> contextFilter;
+
+    private final Semaphore available;
+
+    public LoadSheddingRequestFilter() {
+        this(null);
+    }
+
+    public LoadSheddingRequestFilter(Predicate<RoutingContext> contextFilter) {
+        Optional<String> maxConcurrentRequests = Configuration.getRuntimeProperty("kc.http-max-concurrent-requests");
+        if (maxConcurrentRequests.isPresent()) {
+            available = new Semaphore(Integer.parseInt(maxConcurrentRequests.get()), false);
+        } else {
+            available = null;
+        }
+        this.contextFilter = contextFilter;
+    }
+
+    private final LongAdder rejectedRequests = new LongAdder();
+    private volatile boolean loadSheddingActive;
+
+    @Override
+    public void handle(RoutingContext context) {
+        if (available == null || ignoreContext(context)) {
+            context.next();
+            return;
+        }
+        if (available.tryAcquire()) {
+            context.addEndHandler(event -> {
+                available.release();
+            });
+            if (loadSheddingActive) {
+                // at maximum log one event per second
+                synchronized (rejectedRequests) {
+                    if (loadSheddingActive) {
+                        loadSheddingActive = false;
+                        // rejectedRequests.sumThenReset() is approximative when concurrent increments are active, still it should be accurate enough for this log message
```

Comment:
If the idea is to provide a best effort (not 100% accurate) about the number and a better output in logs, I would avoid using `synchronized` here. This code runs in an event loop thread if we can avoid blocking at all costs, better. 

All that to say I'm not sure about introducing synchronization for logging purposes ...

What if expose these statistics as a metric and potentially introduce a health check for it too?

Reply:
I've moved the logging out of the block. I'd like to add metrics in a second iteration. Waiting for the builds to complete, will then mark it ready-for-review.